### PR TITLE
Specify a single location or a specific region for a custom switch block

### PR DESCRIPTION
**Pull request by saaramahmoudi**
This PR generalize the custom switch block description to allow the user to specify a exact (x,y) location for the switch block by setting type="XY_SPECIFIED", the exact syntax is explained below:

- You should set the switch block type to XY_SPECIFIED (upper case), the other options are: EVERYWHERE, CORNER, ..
- Once you set the type to XY_SPECIFIED, you can put an exact location by setting a "x" and "y" attribute, for example: x="1" y="2"
- If you want to specify a complete row, you can set "x" to -1, for example: x="-1" y="2"
- if you want to specify a complete column, you can set "y" to -1, for example: x="2" y="-1"
- If you want to specify a part of grid, you can specify a region using regular expression documented here: https://mithro-vtr.readthedocs.io/en/latest/arch/reference.html#tag-%3Cregiontype= (please search for word "region" in layout documentation)

#### How Has This Been Tested?
The x and y attribute is tested by @petergrossmann21, and the region description is tested by myself.  

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

I will add a new test and update documentation later today, but I have created the PR to start the review process.

### CODE REVIEW

**Comment by vaughnbetz** [comment]
Thanks @saaramahmoudi . I'll review shortly. Can you also please update the documentation and add it to this PR?

File: vpr/src/route/build_switchblocks.cpp [hunk 1]
```diff
@@ -324,10 +327,11 @@ t_sb_connection_map* alloc_and_load_switchblock_permutations(const t_chan_detail
         if (directionality != sb.directionality) {
             VPR_FATAL_ERROR(VPR_ERROR_ARCH, "alloc_and_load_switchblock_connections: Switchblock %s does not match directionality of architecture\n", sb.name.c_str());
         }
+        
         /* Iterate over the x,y coordinates spanning the FPGA. */
         for (size_t x_coord = 0; x_coord < grid.width(); x_coord++) {
             for (size_t y_coord = 0; y_coord <= grid.height(); y_coord++) {
-                if (sb_not_here(grid, x_coord, y_coord, sb.location)) {
+                if (sb_not_here(grid, x_coord, y_coord, sb.location, sb.x, sb.y, sb.reg_x, sb.reg_y)) {
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:334]: The number of arguments can be reduce by passing sb by reference

File: vpr/src/route/build_switchblocks.cpp [hunk 2]
```diff
@@ -324,10 +327,11 @@ t_sb_connection_map* alloc_and_load_switchblock_permutations(const t_chan_detail
         if (directionality != sb.directionality) {
             VPR_FATAL_ERROR(VPR_ERROR_ARCH, "alloc_and_load_switchblock_connections: Switchblock %s does not match directionality of architecture\n", sb.name.c_str());
         }
+
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:330]: can you please change the for loop that iterates over switchblocks to a range-based for loop that iterate over elements by reference?

Github doesn't allow me to add the comment there (line 323)

File: vpr/src/route/build_switchblocks.cpp [hunk 3]
```diff
@@ -268,7 +268,7 @@ int get_wire_segment_length(const DeviceGrid& grid, e_rr_type chan_type, const t
 static int get_switchpoint_of_wire(const DeviceGrid& grid, e_rr_type chan_type, const t_chan_seg_details& wire_details, int seg_coord, e_side sb_side);
 
 /* returns true if the coordinates x/y do not correspond to the location specified by 'location' */
-static bool sb_not_here(const DeviceGrid& grid, int x, int y, e_sb_location location);
+static bool sb_not_here(const DeviceGrid& grid, int x, int y, e_sb_location location, int sb_x, int sb_y, const e_sb_loc_spec& sb_reg_x, const e_sb_loc_spec& sb_reg_y);
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:271]: replace newly added args with a const t_switchblock_inf& arg

File: vpr/src/route/build_switchblocks.cpp [hunk 4]
```diff
@@ -279,6 +279,9 @@ static bool is_perimeter(const DeviceGrid& grid, int x, int y);
 /* checks if the specified coordinates correspond to the core of the FPGA (i.e. not perimeter) */
 static bool is_core(const DeviceGrid& grid, int x, int y);
 
+/*checks if the specified coordinates matches the exact one specified in the SB architecture description*/
+static bool match_sb_xy(const DeviceGrid& grid, int x, int y, int sb_x, int sb_y, const e_sb_loc_spec& sb_reg_x, const e_sb_loc_spec& sb_reg_y);
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:283]: replace newly added args with a const t_switchblock_inf& arg

File: vpr/src/route/build_switchblocks.cpp [hunk 5]
```diff
@@ -388,6 +392,12 @@ static bool sb_not_here(const DeviceGrid& grid, int x, int y, e_sb_location loca
             if (is_perimeter(grid, x, y) && !is_corner(grid, x, y)) {
                 sb_not_here = false;
             }
+            break;
+        case E_XY_SPECIFIED:
+            if(match_sb_xy(grid, x, y, sb_x, sb_y, sb_reg_x, sb_reg_y)){
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:397]: add space between if and (
between ) and {

File: vpr/src/route/build_switchblocks.cpp [hunk 6]
```diff
@@ -388,6 +392,12 @@ static bool sb_not_here(const DeviceGrid& grid, int x, int y, e_sb_location loca
...
+                sb_not_here = false;
+            }
+
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:400]: remove this empty line to be consistent with the rest of cases

File: vpr/src/route/build_switchblocks.cpp [hunk 7]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
     return is_core;
 }
 
+static bool match_sb_xy(const DeviceGrid& grid, int x, int y, int sb_x, int sb_y, const e_sb_loc_spec& sb_reg_x, const e_sb_loc_spec& sb_reg_y){
+    
+    //if one of sb_x and sb_y is defined, we either know the exact location (x,y) or the exact x location (will apply it to all columns) 
+    //or the exact y location (will apply it to all rows)   
+    if(sb_x != -1 || sb_y != -1){
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:439]: add spaces to all if statements in this function

File: vpr/src/route/build_switchblocks.cpp [hunk 8]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
...
+        
+        if(sb_x == -1 && y == sb_y){
+            return true;
+        }
+    }
+
+    //if both sb_x and sb_y is not defined, we have a region that we should apply this SB pattern to, we just need to check
+    //whether the location passed into this function falls within this region or not
+    if(sb_x == -1 && sb_y == -1){
+        //calculate the appropiate region based on the repeatx/repeaty and current location.
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:456]: appropiate --> appropriate

File: vpr/src/route/build_switchblocks.cpp [hunk 9]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
...
+        //    is not necessary crossponds to the current SB. If incrx/incry is equal to 1, then all locations within the 
+        //    calculated region is valid. 
+        
+        //calculate the region
+        int x_reg_step = (sb_reg_x.repeat != 0) ? (x - sb_reg_x.start) / sb_reg_x.repeat : sb_reg_x.start;
+        int y_reg_step = (sb_reg_y.repeat != 0) ? (y - sb_reg_y.start) / sb_reg_y.repeat : sb_reg_y.start;
+
+        //step must be non-negative
+        x_reg_step = (x_reg_step < 0) ? 0 : x_reg_step;
+        y_reg_step = (y_reg_step < 0) ? 0 : y_reg_step;
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:471]: use std::max(0, value)

File: vpr/src/route/build_switchblocks.cpp (hunk 7 above)
Comment (soheilshahrouz) [reply vpr/src/route/build_switchblocks.cpp:439]: Can't the outer if statement be removed?

File: vpr/src/route/build_switchblocks.cpp [hunk 10]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
     return is_core;
 }
 
+static bool match_sb_xy(const DeviceGrid& grid, int x, int y, int sb_x, int sb_y, const e_sb_loc_spec& sb_reg_x, const e_sb_loc_spec& sb_reg_y){
+    
+    //if one of sb_x and sb_y is defined, we either know the exact location (x,y) or the exact x location (will apply it to all columns) 
+    //or the exact y location (will apply it to all rows)
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:438]: When we know the exact x location while y is not specified, doesn't this mean that we know the column and apply it all the rows?

File: vpr/src/route/build_switchblocks.cpp [hunk 11]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
...
+    //whether the location passed into this function falls within this region or not
+    if(sb_x == -1 && sb_y == -1){
+        //calculate the appropiate region based on the repeatx/repeaty and current location.
+        //This is to determine whether the given location is part of the current SB specified region with regular expression or not
+        //After region calculation, the current SB will apply to this location if:
+        // 1) the given (x,y) location falls into the calculated region 
+        // *AND*
+        // 2) incrx/incry are respected within the region, this means all locations within the calculated region is
+        //    is not necessary crossponds to the current SB. If incrx/incry is equal to 1, then all locations within the 
+        //    calculated region is valid.
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:463]: a few typos
is is --> do
necessary --> necessarily
crossponds -- > correspond
is valid --> are valid

File: vpr/src/route/build_switchblocks.cpp [hunk 12]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
...
+
+        int reg_starty = sb_reg_y.start + (y_reg_step * sb_reg_y.repeat);
+        int reg_endy = sb_reg_y.end + (y_reg_step * sb_reg_y.repeat);
+        reg_endy = std::min(reg_endy, int(grid.height() - 1));
+  
+        //check x coordinate
+        if (x >= reg_startx && x <= reg_endx){ //should fall into the region
+            //we also should respect the incrx
+            //if incrx is not equal to 1, all locations within this region is *NOT* valid
+            if((y + reg_startx) % sb_reg_x.incr == 0){
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:485]: shouldn't this be x + reg_startx?

File: vpr/src/route/build_switchblocks.cpp [hunk 13]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
...
+        int reg_startx = sb_reg_x.start + (x_reg_step * sb_reg_x.repeat);
+        int reg_endx = sb_reg_x.end + (x_reg_step * sb_reg_x.repeat);
+        reg_endx = std::min(reg_endx, int(grid.width() - 1));
+
+        int reg_starty = sb_reg_y.start + (y_reg_step * sb_reg_y.repeat);
+        int reg_endy = sb_reg_y.end + (y_reg_step * sb_reg_y.repeat);
+        reg_endy = std::min(reg_endy, int(grid.height() - 1));
+  
+        //check x coordinate
+        if (x >= reg_startx && x <= reg_endx){ //should fall into the region
```
Comment (soheilshahrouz) [inline vpr/src/route/build_switchblocks.cpp:482]: these nested if statements are a little hard to read
I think the whole check can be implemented in a single if statement

File: libs/libarchfpga/src/read_xml_arch_file.cpp [hunk 14]
```diff
@@ -3932,11 +3942,72 @@ static void ProcessSwitchblocks(pugi::xml_node Parent, t_arch* arch, const pugiu
...
+
+                //parse the values from the architecture file and fill out SB region information
+                vtr::FormulaParser p;
+
+                vtr::t_formula_data vars;
+                vars.set_var_value("W", grid_width);
+                vars.set_var_value("H", grid_height);
+
+                
+                sb.reg_x.start = startx_attr.empty() ? 0 : p.parse_formula(startx_attr.value(), vars);
```
Comment (soheilshahrouz) [inline libs/libarchfpga/src/read_xml_arch_file.cpp:3995]: Other functions in this file do not seem to parse formulas when reading the architecture file. Instead they store formulas in std:string member variables and parse them later.

**Review by petergrossmann21 — CHANGES_REQUESTED** [review]
@saaramahmoudi Can you include the tests that verify this feature in the PR?

File: vpr/src/route/build_switchblocks.cpp [hunk 15]
```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
...
+        //This is to determine whether the given location is part of the current SB specified region with regular expression or not
+        //After region calculation, the current SB will apply to this location if:
+        // 1) the given (x,y) location falls into the calculated region 
+        // *AND*
+        // 2) incrx/incry are respected within the region, this means all locations within the calculated region is
+        //    is not necessary crossponds to the current SB. If incrx/incry is equal to 1, then all locations within the 
+        //    calculated region is valid. 
+        
+        //calculate the region
+        int x_reg_step = (sb_reg_x.repeat != 0) ? (x - sb_reg_x.start) / sb_reg_x.repeat : sb_reg_x.start;
```
Comment (vaughnbetz) [inline vpr/src/route/build_switchblocks.cpp:466]: I'm not following the logic very clearly here. Clearly defining what .repeat and .start mean in the structure definition would help, as would a bit more of an overview of the logic here.

File: vpr/src/route/build_switchblocks.cpp (hunk 12 above)
Comment (vaughnbetz) [reply vpr/src/route/build_switchblocks.cpp:485]: Agreed.

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
Some comments to address (mine partially overlap Soheil's).
Also, it would be good to make the documentation update now, as part of this PR as well. That also might help clarify the precise meaning of repeat and increment (as would comments on the structure data members storing the sb_loc information).

**Comment by petergrossmann21** [comment]
@saaramahmoudi I don't see any new files related to testing; is there something that still needs to be added?

**Comment by saaramahmoudi** [comment]
@petergrossmann21 @vaughnbetz I am so sorry, but I have been busy doing a different task at work. I am trying my best to land this PR as soon as possible. It might take two-three weeks.

**Comment by vaughnbetz** [comment]
@saaramahmoudi : I think this is ready to merge, but I just merged a libcatch2 update that caused a conflict (sorry!).  If you fix that conflict I think this is good to go.

**Comment by saaramahmoudi** [comment]
@vaughnbetz 
Here is the link to Titan_quick_qor benchmarks: [Titan](https://docs.google.com/spreadsheets/d/10fMlD4YQVFGoVXUgq9RRG808WQGaqn5v34SFwhJWges/edit?usp=sharing)

**Comment by vaughnbetz** [comment]
QoR looks good, thanks.

**Comment by saaramahmoudi** [comment]
@vaughnbetz This branch is also ready to merge

**Comment by vaughnbetz** [comment]
Thanks @saaramahmoudi !

**Pull request merged**